File: src/nfan/fan/sys/Obj.cs
//
// Copyright (c) 2006, Brian Frank and Andy Frank
// Licensed under the Academic Free License version 3.0
//
// History:
//   14 Sep 06  Andy Frank  Creation
//

using System;

namespace Fan.Sys
{
  ///
  /// Obj is the root type for all classes and mixins in Fan.  It's an interface
  /// so that it can handle mixin interfaces.  The root class is FanObj.
  ///
  public interface Obj
  {

    Bool equals(Obj obj);

    Int compare(Obj obj);

    Int hash();

    Str toStr();

    Bool isImmutable();

    Type type();

    Obj trap(Str name, List args);

  }
}

File: src/sys/es/fan/Obj.js
//
// Copyright (c) 2008, Brian Frank and Andy Frank
// Licensed under the Academic Free License version 3.0
//
// History:
//   11 Dec 2008  Andy Frank  Creation
//   20 May 2009  Andy Frank  Refactor to new OO model
//   31 Mar 2023  Matthew Giannini  Refactor to ES
//

/**
 * Obj
 */
class Obj {

  /** Simple counter-based hash for object that do not override hash() */
  static #hashCounter = 0;
  #hash;

  equals(that) { return this === that; }

  compare(that) {
    if (this < that) return -1;
    if (this > that) return 1;
    return 0;
  }

  hash() {
    if (this.#hash === undefined) this.#hash = Obj.#hashCounter++;
    return this.#hash;
  }

  with$(f) {
    f(this);
    return this;
  }

  isImmutable() { return this.typeof$().isConst(); }

  toImmutable() {
    if (this.isImmutable()) return this;
    throw NotImmutableErr.make(this.typeof$().toStr());
  }

  toStr() { return `${this.typeof$()}@${this.hash()}`; }

  toString() { return "" + this.toStr(); }

  trap(name, args=null) { return ObjUtil.doTrap(this, name, args, this.typeof$()); }
}
